src/python_testing/TC_IDM_2_2.py

```diff
@@ -0,0 +1,674 @@
+
+        elif desired_type == IntFlag:
+            if hasattr(cluster, 'Attributes'):
+                attributes_class = getattr(cluster, 'Attributes')
+                if hasattr(attributes_class, 'FeatureMap'):
+                    all_attributes_of_type = [cluster.Attributes.FeatureMap]
+                else:
+                    raise Exception(f'Cluster {cluster} lacks a FeatureMap')
+            else:
+                raise Exception(f'Cluster {cluster} lacks attributes')
+        else:
+            all_attributes_of_type = [attribute for attribute in all_attributes if attribute.attribute_type ==
+                                      ClusterObjects.ClusterObjectFieldDescriptor(Type=desired_type)]
+
+        return all_attributes_of_type
+
+    def all_device_clusters(self) -> set:
+        device_clusters = set()
+        for endpoint in self.endpoints:
+            device_clusters |= set(self.endpoints[endpoint].keys())
+        return device_clusters
+
+    def all_device_attributes(self) -> set:
+        device_attributes = set()
+        for endpoint in self.endpoints:
+            for cluster in self.endpoints[endpoint]:
+                device_attributes |= self.endpoints[endpoint][cluster].keys()
+        return device_attributes
+
+    def verify_attribute_list_cluster(self, read_request: dict, endpoint: int, cluster: ClusterObjects.Cluster):
+        cluster_ids = [c.id for c in read_request[endpoint].keys()]
+        asserts.assert_equal(
+            sorted(read_request[endpoint][Clusters.Descriptor][Clusters.Descriptor.Attributes.ServerList]),
+            sorted(cluster_ids),
+            f"ServerList doesn't match the expected server list for endpoint {endpoint}"
+        )
+
+    async def check_attribute_read_for_type(self, desired_attribute_type: type) -> bool:
+        # Get all clusters from device
+        for cluster in self.device_clusters:
+            all_types = await self.all_type_attributes_for_cluster(cluster, desired_attribute_type)
+            all_types = list(set(all_types) & self.device_attributes)
+            if all_types:
+                chosen_attributes = all_types
+                cluster = Clusters.ClusterObjects.ALL_CLUSTERS[chosen_attributes[0].cluster_id]
+                break
+        else:
+            print(f"Attribute not found on device: {desired_attribute_type}")
+            cluster = None
+
+        endpoint = None
+        for endpoint in self.endpoints:
+            for cluster, clusterdata in self.endpoints[endpoint].items():
+                all_types = await self.all_type_attributes_for_cluster(cluster, desired_attribute_type)
+                attributes_of_type = set(all_types)
+                attributes_of_type_on_device = attributes_of_type.intersection(set(clusterdata.keys()))
+                if attributes_of_type_on_device:
+                    return True
+        return False
+
+    async def read_unsupported_attribute(self):
```

Comment:
this function is called read unsupported, but it doesn't actually read unsupported. I don't think this function does anything - there's no return value and no actions are performed.

Reply:
There is a self.read_single_attribute_expect_error call in this method, so I am confused as to what you mean.